### PR TITLE
Reinstate tcp reconnect task

### PR DESCRIPTION
**Pull request by simpsont-oci**

### CODE REVIEW

File: dds/DCPS/transport/tcp/TcpConnection.h [hunk 1]
```diff
@@ -206,7 +207,8 @@ class TcpConnection
 
   /// Small unique identifying value.
   std::size_t id_;
-  ACE_thread_t reconnect_thread_;
+  //ACE_thread_t reconnect_thread_;
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.h]:
```suggestion
```

File: dds/DCPS/transport/tcp/TcpReconnectTask.h [hunk 2]
```diff
@@ -21,35 +23,33 @@ OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
...
+class TcpReconnectTask : public ACE_Task_Base {
 public:
-  TcpReconnectTask(TcpConnection* con);
+  TcpReconnectTask();
 
   virtual ~TcpReconnectTask();
 
-  /// Handle reconnect requests.
-  virtual void execute(ReconnectOpType& op);
+  bool reconnect(TcpConnection_rch con);
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpReconnectTask.h]: Should this return type be `void`?  I don't see it used for anything.

File: dds/DCPS/ReactorInterceptor.h [hunk 3]
```diff
@@ -29,7 +29,7 @@ class OpenDDS_Dcps_Export ReactorInterceptor : public RcEventHandler {
   class Command
   : public RcObject {
   public:
-    Command() : executed_(false), condition_(mutex_) {}
+    Command() : executed_(false), condition_(mutex_), reactor_(NULL) {}
```
Comment (mitza-oci) [inline dds/DCPS/ReactorInterceptor.h]:
```suggestion
    Command() : executed_(false), condition_(mutex_), reactor_(0) {}
```

File: dds/DCPS/ReactorInterceptor.h [hunk 4]
```diff
@@ -54,13 +54,32 @@ class OpenDDS_Dcps_Export ReactorInterceptor : public RcEventHandler {
...
-    ACE_Thread_Mutex mutex_;
-    ACE_Condition_Thread_Mutex condition_;
+    mutable ACE_Thread_Mutex mutex_;
+    mutable ACE_Condition_Thread_Mutex condition_;
+    ACE_Reactor* reactor_;
   };
   typedef RcHandle<Command> CommandPtr;
 
+  template <typename T>
+  class ResultCommand : public ReactorInterceptor::Command {
```
Comment (mitza-oci) [inline dds/DCPS/ReactorInterceptor.h]:
```suggestion
  class ResultCommand : public Command {
```

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 5]
```diff
@@ -77,24 +76,44 @@ OpenDDS::DCPS::TcpConnection::TcpConnection(const ACE_INET_Addr& remote_address,
...
 }
 
+void
+OpenDDS::DCPS::TcpConnection::set_datalink(const OpenDDS::DCPS::TcpDataLink_rch& link)
+{
+  DBG_ENTRY_LVL("TcpConnection","set_datalink",6);
+
+  GuardType guard(reconnect_lock_);
+
+  // Keep a "copy" of the reference to the data link for ourselves.
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]:
```suggestion
```
I realize this comment line came from the code before (in .inl) but let's just remove it.  We can assume the reader is acquainted with the basics of reference counting.

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 6]
```diff
@@ -837,8 +858,15 @@ void
 OpenDDS::DCPS::TcpConnection::transfer(TcpConnection* connection)
 {
   DBG_ENTRY_LVL("TcpConnection","transfer",6);
+  GuardType guard(reconnect_lock_);
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]: Looks like a candidate for refactoring with common code in `shutdown`.

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 7]
```diff
@@ -630,19 +648,17 @@ OpenDDS::DCPS::TcpConnection::passive_reconnect_i()
...
-
-      if (this->passive_reconnect_timer_id_ == -1) {
-        ACE_ERROR_RETURN((LM_ERROR,
-                          ACE_TEXT("(%P|%t) ERROR: TcpConnection::passive_reconnect_i")
-                          ACE_TEXT(", %p.\n"), ACE_TEXT("schedule_timer")),
-                         -1);
+    if (interceptor_ && passive_reconnect_timer_id_ == -1) {
+      ReactorInterceptor::CommandPtr cmd = interceptor_->execute_or_enqueue(new ScheduleTimer(RcHandle<TcpConnection>(this, inc_count()), 0, timeout));
+      guard.release();
+      long result = static_rchandle_cast<ReactorInterceptor::ResultCommand<long> >(cmd)->wait_result();
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]: I think we could avoid the cast here by providing a static member of `ResultCommand<T>`:
```
  static T wait_result(CommandPtr cmd);
```
To make the calling code: `const long result = ReactorInterceptor::ResultCommand<long>::wait_result(cmd);`

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 8]
```diff
@@ -837,8 +858,15 @@ void
...
   DBG_ENTRY_LVL("TcpConnection","transfer",6);
+  GuardType guard(reconnect_lock_);
+  ACE_Reverse_Lock<ACE_Thread_Mutex> rev_lock(reconnect_lock_);
+  while (reconnect_task_->active()) {
+    ACE_GUARD(ACE_Reverse_Lock<ACE_Thread_Mutex>, rev_guard, rev_lock);
+    reconnect_task_->wait_complete();
+  }
 
-  GuardType guard(this->reconnect_lock_);
+  if (shutdown_)
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]:
```suggestion
  if (shutdown_) {
```

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 9]
```diff
@@ -112,18 +131,17 @@ OpenDDS::DCPS::TcpConnection::disconnect()
 {
   DBG_ENTRY_LVL("TcpConnection","disconnect",6);
   this->connected_ = false;
-  TcpReceiveStrategy_rch receive_strategy = this->receive_strategy();
-  if (receive_strategy) {
-    receive_strategy->get_reactor()->remove_handler(this,
-                                                    ACE_Event_Handler::READ_MASK | ACE_Event_Handler::DONT_CALL);
+
+  if (interceptor_) {
+    ReactorInterceptor::CommandPtr cmd = interceptor_->execute_or_enqueue(new RemoveHandler(RcHandle<TcpConnection>(this, inc_count()), READ_MASK | DONT_CALL));
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]: There's a lot going on here.  Maybe break out the `TcpConnection_rch` construction to its own automatic variable. (same with other enqueue_or_execute calls below)

File: dds/DCPS/transport/tcp/TcpReconnectTask.cpp [hunk 10]
```diff
@@ -6,48 +6,93 @@
...
+
+bool TcpReconnectTask::active() {
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  return connection_;
+}
+
+void TcpReconnectTask::shutdown() {
+  DBG_ENTRY_LVL("TcpReconnectTask", "shutdown", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  shutdown_ = true;
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpReconnectTask.cpp]: Should this wake up a `wait_complete()`?

File: dds/DCPS/transport/tcp/TcpReconnectTask.cpp [hunk 11]
```diff
@@ -6,48 +6,93 @@
...
+      connection->tear_link();
     }
+  }
+
+  connection_.reset();
+  cv_.broadcast();
 
-  } else
-    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: TcpReconnectTask::svc unknown operation %d\n", op));
+  g.release();
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpReconnectTask.cpp]: Is this needed?

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 12]
```diff
@@ -77,24 +76,44 @@ OpenDDS::DCPS::TcpConnection::TcpConnection(const ACE_INET_Addr& remote_address,
...
+OpenDDS::DCPS::TcpConnection::set_datalink(const OpenDDS::DCPS::TcpDataLink_rch& link)
+{
+  DBG_ENTRY_LVL("TcpConnection","set_datalink",6);
+
+  GuardType guard(reconnect_lock_);
+
+  // Keep a "copy" of the reference to the data link for ourselves.
+  link_ = link;
+  if (link_) {
+    impl_ = RcHandle<TcpTransport>(static_cast<TcpTransport*>(&link_->impl()), inc_count());
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]:
```suggestion
    impl_ = TcpTransport_rch(static_cast<TcpTransport*>(&link_->impl()), inc_count());
```

There are some similar cases below -- use the *_rch typedef for consistency with the rest of the code

File: dds/DCPS/transport/tcp/TcpConnection.cpp [hunk 13]
```diff
@@ -77,24 +76,44 @@ OpenDDS::DCPS::TcpConnection::TcpConnection(const ACE_INET_Addr& remote_address,
...
+    if (impl_) {
+      interceptor_ = make_rch<Interceptor>(impl_->reactor_task().get(), impl_->reactor_task()->get_reactor(), impl_->reactor_task()->get_reactor_owner());
```
Comment (mitza-oci) [inline dds/DCPS/transport/tcp/TcpConnection.cpp]: Extract the `reactor_task` as its own variable?

File: dds/DCPS/transport/tcp/TcpReconnectTask.cpp (hunk 11 above)
Comment (simpsont-oci) [reply dds/DCPS/transport/tcp/TcpReconnectTask.cpp]: Yeah, if we happen to be the last reference holder for TcpConnection, we need to allow it to call back in to TcpReconnectTask::shutdown().

File: dds/DCPS/transport/tcp/TcpReconnectTask.cpp (hunk 10 above)
Comment (simpsont-oci) [reply dds/DCPS/transport/tcp/TcpReconnectTask.cpp]: The goal here was to have connection_ act as the single indicator for whether work was scheduled to be done, and so even if shutdown() has been called & shutdown_ is set, wait_complete would still want to wait until connection_ is no longer set.

**Pull request merged**